x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/BinaryDvConfirmedAutomatonQuery.java

```diff
@@ -27,32 +30,65 @@
- * Expensive to run so normally used in conjunction with more selective query clauses.
+ * Query that runs an Automaton across all binary doc values (but only for docs that also 
+ * match a provided approximation query which is key to getting good performance).
  */
-public class AutomatonQueryOnBinaryDv extends Query {
+public class BinaryDvConfirmedAutomatonQuery extends Query {
 
     private final String field;
     private final String matchPattern;
     private final ByteRunAutomaton bytesMatcher;
+    private final Query approxQuery;
 
-    public AutomatonQueryOnBinaryDv(String field, String matchPattern, Automaton automaton) {
+    public BinaryDvConfirmedAutomatonQuery(Query approximation, String field, String matchPattern, Automaton automaton) {
+        this.approxQuery = approximation;
         this.field = field;
         this.matchPattern = matchPattern;
         bytesMatcher = new ByteRunAutomaton(automaton);
     }
+    public BinaryDvConfirmedAutomatonQuery(String field, String matchPattern, Automaton automaton) {
+        this(new MatchAllDocsQuery(), field, matchPattern, automaton);
+    }
+    
+    private BinaryDvConfirmedAutomatonQuery(Query approximation, String field, String matchPattern, ByteRunAutomaton bytesMatcher) {
+        this.approxQuery = approximation;
+        this.field = field;
+        this.matchPattern = matchPattern;
+        this.bytesMatcher = bytesMatcher;
+    }    
+    
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        Query approxRewrite = approxQuery.rewrite(reader);
+        if (approxQuery != approxRewrite ) {
+            return new BinaryDvConfirmedAutomatonQuery(approxRewrite, field, matchPattern, bytesMatcher);
+        }
+        return this;
+    }    
 
     @Override
     public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        final Weight approxWeight = approxQuery.createWeight(searcher, scoreMode, boost);
+        
         return new ConstantScoreWeight(this, boost) {
 
             @Override
             public Scorer scorer(LeafReaderContext context) throws IOException {
                 ByteArrayDataInput badi = new ByteArrayDataInput();
                 final BinaryDocValues values = DocValues.getBinary(context.reader(), field);
-                TwoPhaseIterator twoPhase = new TwoPhaseIterator(values) {
+                Scorer approxScorer = approxWeight.scorer(context);
+                if (approxScorer == null) {
+                    // No matches to be had
+                    return null;
+                }
+                DocIdSetIterator approxDisi = approxScorer.iterator();
+                TwoPhaseIterator twoPhase = new TwoPhaseIterator(approxDisi) {
                     @Override
                     public boolean matches() throws IOException {
+                        if (values.advanceExact(approxDisi.docID()) == false)
```

Comment:
Would it make more sense to assert that `advanceExact` always returns true? It seems like it'd indicate a serious bug in our logic if this returned false.

Reply:
Will add an assertion. We'll still need the if for when running without assertions on.